### PR TITLE
Created South Africa preprocessing script

### LINKED ISSUE #41: Develop preprocessing script for legislation from South Africa

### PR DESCRIPTION
**Pull request by Skyrolon**
Closes #41

### CODE REVIEW

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
...tural beauty as well as the preservation of biotic diversity in the natural environment; [b] .the promotion of sustained utilization of species and ecosystems and the effective application and re-use of natural resources; [c] the protection of the environment against disturbance, deterioration, defacement, poisoning or destruction as a result of man-made structures, installations, processes or products or human activities; and [d] the establishment, maintenance and improvement of environments which contribute to a generally acceptable quality of life for the inhabitants of the Republic of South Africa. [2] The policy contemplated in subsection [1] shall be determined by the Minister after consultation with the council and the Administrator of each province, and with the concurrence of: [a] each Minister charged with the administration of any law which in the opinion of the Minister relates to a matter affecting the environment; and [b] the Minister of Finance and the Minister of Economic Affairs and Technology. [3] The Minister may at any time, subject to the provisions of subsection [2], by like notice substitute, withdraw or amend the policy determined in terms of subsection [1].
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `[b] .the` with `[b] the`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+Committees of council
+(7) [1][a] The council may with the approval of the Minister from among its number elect an executive committee consisting of such number of members as the council may determine with the approval of the Minister. [b] The chairman of the council shall be the chairman of the executive committee. [c] The executive committee shall perform such functions as the council may determine with the approval of the Minister. [2][a] The council may with the approval of the Minister establish other committees consisting of such persons as may be determined by the council in order to assist the council in the performance of its functions. [b] The council shall designate a member of a committee as chairman of such committee. [3] The council may allow any member of a committee who is not a member of the council to attend any meeting of the council at which any matter relating to a function assigned to that committee is dealt with and may allow such member to take part in the proceedings at such meeting, without having the right to vote.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `[1][a]` with `[1] [a]`. Also, please replace `[2][a]` with `[2] [a]`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+Administrative work of council
+(10) The Director-General shall designate, subject to the provisions of the Public Service Act, 1984 (Act No. 111 of 1984), as many officers and employees of the Department as may be necessary to assist the council and any committee thereof in the administrative work connected with the performance of the functions of such council or committee: Provided that, with the approval of the Minister, such administrative work may be performed by any person other than such officer or employee at the remuneration and allowances which the Minister with the concurrence of the Minister of Finance may determine.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `committee: Provided` with `committee, provided`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
...tion referred to in subsection [4][a] to appear before the board of investigation shall be by way of a subpoena signed by the chairman of the board. [6] [a] A session of the board of investigation shall be held in public. [b] The decision of the board and the reasons therefor shall be reduced to writing. [7] A member of the board of investigation who is not in the full-time employment of the State may be paid from money appropriated by Parliament for that purpose such remuneration and allowances as the Minister may, with the concurrence of the Minister of Finance, determine either in general or in any particular case. [8] The Director-General shall designate, subject to the provisions of the Public Service Act, 1984 (Act No. 111 of 1984), as many officers and employees of the Department as may be necessary to assist the board in the administrative work connected with the performance of the functions of the board of investigation: Provided that with the approval of the Minister such administrative work may be performed by any person other than such officer or employee at the remuneration and allowances which the Minister with the concurrence of the Minister of Finance may determine.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `(i)` with `[i]`. Also, please replace ` arid ` with ` and `. Furthermore, please replace `investigation: Provided` with `investigation, provided`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
...ccessors in title of such owner and holder of the real right, shall be subject to the provisions of such directions. [4] The Administrator shall in writing direct the registrar of deeds of the deeds registry in which the title deed of land referred to in subsection [3] is registered, to make an entry of the directions in question in his registers and to endorse the office copy of the title deed accordingly. [5] The Administrator may with the concurrence of the Minister of Finance out of money appropriated by Parliament for that purpose and subject to such conditions as he may determine, render financial aid by way of grants or otherwise to the owner of, and the holder of a real right in, land situated within a protected natural environment in respect of expenses incurred by the owner or holder of the right in compliance with any direction issued in terms of subsection [2]. [6] An Administrator may: [a] with the concurrence of a local authority or government institution assign the control and management of a protected natural environment to such local authority or government institution; or [b] withdraw such control and management from such local authority or government institution.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace both instances of `area: Provided` with `area, provided`. Also, please replace `Act: Provided` with `Act, provided`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
...necessary to assist a management advisory committee in the administrative work connected with the performance of the functions of the committee: Provided that where the control and management of a protected natural environment has been assigned to another local authority or government institution in terms of section 16[6] the chief executive officer of such local authority or government institution shall designate as many employees of the relevant local authority or government institution as may be necessary to assist a management advisory committee with the said administrative work: Provided further that with the approval of the Minister such administrative work may be performed by any person other than such officer or employee at the remuneration and allowances which the Minister with the concurrence of the Minister of Finance may determine. [8] A member of a management advisory committee who is not in the full-time employment of the State or a local authority may be paid from money appropriated by Parliament for that purpose, such remuneration and allowances as the Administrator may, with the concurrence of the Minister of Finance, determine in general or in any particular case.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `office: Provided` with `office, provided`. Also, please replace `committee: Provided` with `committee, provided`. Furthermore, please replace `work: Provided` with `work, provided`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
...shall be in the form and be accompanied by such information as the Minister may prescribe. [3] If the Minister of Water Affairs should require any further information to enable him to make a decision on an application for a permit referred to in subsection [1], he may demand such information from the applicant. [4] The Minister of Water Affairs shall maintain a register in which details of every disposal site for which a permit has been issued shall be recorded." [5] The Minister of Water Affairs may from time to time by notice in the Gazette issue directions with regard to: [a] the control and management of disposal sites in general; [b] the control and management of certain disposal sites or disposal sites handling particular types of waste; and [c] the procedure to be followed before any disposal site may be withdrawn from use or utilized for another purpose. [6] Subject to the provisions of any other law no person shall discard waste or dispose of it in any other manner, except: [a] at a disposal site for which a permit has been issued in terms of subsection [1]; or [b] in a manner or by means of a facility or method and subject to such conditions as the Minister may prescribe.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `recorded."` with `recorded.`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+PARTY CONTROL OF ACTIVITIES WHICH MAY HAVE DETRIMENTAL EFFECT ON THE ENVIRONMENT
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `PARTY` with `PART V -`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+Identification of activities which will probably have detrimental effect on environment
+(21) [1] The Minister may by notice in the Gazette identify those activities which in his opinion may have a substantial detrimental effect on the environment. whether in general or in respect of certain areas. [2] Activities which are identified in terms of subsection [1] may include any activity in any of the following categories., but are not limited thereto: [a] Land use and transformation; [b] water use and disposal; [c] resource removal, including natural living resources; [d] resource renewal; [e] agricultural processes; [f] industrial processes; [g] transportation; [h] energy generation and distribution; [i] waste and sewage disposal; [j] chemical treatment; [k] recreation. [3] The Minister identifies an activity in terms of subsection [1] after consultation with the council and the Administrator of each province, and with the concurrence of: [a] the Minister of each department of State responsible for the execution, approval or control of such activity; and [b] the Minister of Finance and the Minister of Economic Affairs and Technology.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `environment. whether` with `environment, whether`. Also, please replace `categories.,` with `categories,`. Furthermore, please replace `[a] Land` with `[a] land`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
...lopment or activity to be undertaken unless he has on application been authorized thereto by the Minister or a local authority designated by the Minister in the notice, on the conditions contained in such authorization. [3] In considering an application for an authorization referred to in subsection [2] the Minister or the designated local authority may request the person to submit a report as prescribed, concerning the influence of the proposed activity on the environment in the limited development area. [4] A limited development area shall not be declared unless the Minister: [a] has given notice in the Gazette and in not fewer than one English and one Afrikaans newspaper circulating in the area in question of his intention to declare such area as a limited development area; [b] has permitted not fewer than  days for the submission to the Director General of comment on the proposed declaration; [c] has considered all representations received in terms of such notice; and [d] has consulted each Minister charged with the administration of any law which in the opinion of the Minister relates to a matter affecting the environment in that area, as well as the Administrator in question.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `not fewer than  days` with `not fewer than 60 days`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
...ine and such imprisonment. [6] Any person convicted of an offence in terms of this Act, and who after such conviction persists in the act or omission which constituted such offence, shall be guilty of a continuing offence and liable on conviction to a fine not exceeding R250 or to imprisonment for a period not exceeding 20 days or to both such fine and such imprisonment in respect of every day on which he so persists with such act or omission. [7] In the event of a conviction in terms of this Act the court may order that any damage to the environment resulting from the offence be repaired by the person so convicted, to the satisfaction of the Minister or the local authority concerned. [8] If within a period of days after a conviction or such longer period as the court may determine at the time of the conviction, an order in terms of subsection [7] is not being complied with, the Minister or local authority concerned may itself take the necessary steps to repair the damage and recover the cost thereof from the person so convicted. [9] Notwithstanding anything to the contrary in any law contained, a magistrate's court shall be competent to impose any penalty provided for in this Act.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `who. having` with `who, having`. Also, please replace `Rl,000` with `R1,000`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+Forfeiture 50
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please remove the trailing ` 50`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+Compensation for loss
+(34) [1] If in terms of the provisions of this Act limitations are placed on the purposes for which land may be used or on activities which may be undertaken on the !and, the owner of, and the holder of a real right in, such land shall have a right to recover compensation from the Minister or Administrator concerned in respect of actual loss suffered by him consequent upon the application of such limitations. [2] The amount so recoverable shall be determined by agreement entered into between such owner or holder of the real right and the Minister or Administrator, as the case may be, with the concurrence of the Minister of Finance. [3] In the absence of such agreement the amount so to be paid shall be determined by a court referred to in section 14 of the Expropriation Act, 1975 (Act No. 63 of 1975), and the provisions of that section and section of that Act shall mutatis mutandis apply in determining such amount.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `!and` with `land`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
... of Water Affairs in the prescribed manner, within the prescribed period and upon payment of the prescribed fee. [2] Any person who feels aggrieved at a decision of an officer or employee enforcing a provision of this Act in respect of a protected natural environment may appeal against such decision to the Administrator concerned, in the prescribed manner. within the prescribed period and upon payment of the prescribed fee. [3] Subject to the provisions of subsections [1] and [2] any person who feels aggrieved at a decision of an officer or employee exercising any power delegated to him in terms of this Act or conferred upon him by regulation, may appeal against such decision to the Minister in the prescribed manner, within the prescribed period and upon payment of the prescribed fee. [4] The Minister, the Minister of Water Affairs or an Administrator, as the case may be, may, after considering such an appeal, confirm, set aside or vary the decision of the officer or employee or make such order as he may deem fit, including an order that the prescribed fee paid by the applicant or such part thereof as the Minister or Administrator concerned may determine be refunded to that person.
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please replace `manner. within` with `manner, within`.

File: raw_data/preprocessed/africa/south_africa_english.txt [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+(40) The provisions of this Act shall bind the State, including any provincial administration, except in so far as criminal liability is concerned.
+
+Application of Act
+(41) [1] This Act shall also apply in respect of the Prince Edward Islands as defined in section 1 of the Prince Edward Islands Act, 1948 (Act No. 43 of 1948). [2] The provisions of this Act shall not apply in respect of any matter to which the provisions of the National Monuments Act, 1969 (Act No. 28 of 1969), apply.
+
+Repeal of laws, and savings
+(42) [1] Subject to the provisions of subsection [2], the laws mentioned in the Schedule are hereby repealed to the extent set out in the third column thereof. [2] Anything done under any provision of a law repealed by subsection [1] and which could have been done under a provision of this Act shall be deemed to have been done under the latter provision. 
+
+Amendment of section 1 of Act 88 of 1967, as substituted by section 1 of Act 73 of 1975 and amended by section 1 of Act 104 of 1977, section 1 of Act 51 of 1981, section 1 of Act 87 of 1983 and section 1 of Act 92 of 1985
+
```
Comment (danpdraper) [inline raw_data/preprocessed/africa/south_africa_english.txt]: Please delete this line; the line above is the title of the article below.

File: scripts/legislation/africa/south_africa_english.sh [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+    replace_parentheses_around_article_delimiters_with_square_brackets | \
+    move_article_titles_after_article_numbers | \
+    apply_common_transformations_to_stdin "$language" | \
+    move_article_titles_above_articles | \
+    remove_line_numbers | \
+    amend_errors_in_articles | \
+    amend_errors_in_headers 
+}
+
+#/Users/matthkli/projects/eia/scripts/legislation/preprocess_legislation.sh south_africa english && cat /Users/matthkli/projects/eia/raw_data/preprocessed/africa/south_africa_english.txt
```
Comment (danpdraper) [inline scripts/legislation/africa/south_africa_english.sh]: Please delete these two lines.

File: scripts/legislation/africa/south_africa_english.sh [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+  sed -E 's/^(\([0-9]+\) )([^|]+)\|/\2\n\1/'
+}
+
+function remove_line_numbers {
+  sed -E 's/ 5 | 10 | 15 | 20 | 25 | 30 | 35 | 40 | 45 | 50 | 55 | 60/ /g' 
+}
+
+function amend_errors_in_articles {
+  sed -E 's/-- /: /g' | \
+    sed -E 's/- /: /g' | \
```
Comment (danpdraper) [inline scripts/legislation/africa/south_africa_english.sh]: You should be able to replace these two commands with:

```
sed -E 's/-+ /: /g'
```

File: scripts/legislation/africa/south_africa_english.sh [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+function remove_all_text_before_first_header {
+  sed -n '/^I. In this Act/,$p'
+}
+
+function prefix_article_numbers_with_article_literal {
+  sed -E 's/^([0-9]+\.)/Article \1/'
+}
+
+function replace_parentheses_around_article_delimiters_with_square_brackets {
+  sed -E 's/^\(([A-Za-z0-9]+)\)/[\1]/'
```
Comment (danpdraper) [inline scripts/legislation/africa/south_africa_english.sh]: Due to the `^` character in the regular expression, the expression will only match numbers and letters in parentheses at the beginning of a line. Is that intended?

File: scripts/legislation/africa/south_africa_english.sh [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+function remove_line_numbers {
+  sed -E 's/ 5 | 10 | 15 | 20 | 25 | 30 | 35 | 40 | 45 | 50 | 55 | 60/ /g' 
+}
+
+function amend_errors_in_articles {
+  sed -E 's/-- /: /g' | \
+    sed -E 's/- /: /g' | \
+    sed -E 's/ \(([0-9]+)\)/ [\1]/g' | \
```
Comment (danpdraper) [inline scripts/legislation/africa/south_africa_english.sh]: Isn't the `replace_parentheses_around_article_delimiters_with_square_brackets` function supposed to do this?

File: scripts/legislation/africa/south_africa_english.sh [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+    sed -E 's/([0-9]+) (\[[0-9]+\])/\1\2/g' | \
+    sed -E 's/ \(a\)/ \[a\]/g' | \
```
Comment (danpdraper) [inline scripts/legislation/africa/south_africa_english.sh]: Isn't the `replace_parentheses_around_article_delimiters_with_square_brackets` function supposed to do this?

File: scripts/legislation/africa/south_africa_english.sh [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+    sed -E 's/subsection \(([0-9]+)\)/subsection [\1]/g' | \
```
Comment (danpdraper) [inline scripts/legislation/africa/south_africa_english.sh]: Isn't the `replace_parentheses_around_article_delimiters_with_square_brackets` function supposed to do this?

**Review by danpdraper — CHANGES_REQUESTED** [review]

File: scripts/legislation/africa/south_africa_english.sh (hunk 18 above)
Comment (Skyrolon) [reply scripts/legislation/africa/south_africa_english.sh]: No, thanks for the catch. I have amended the rest of the script to adjust for this change.

**Pull request merged**